benches/common.rs

```diff
@@ -0,0 +1,134 @@
+use std::{any::type_name, convert::TryFrom, fmt::Debug, iter, marker::PhantomData};
+use tango_bench::{
+    BenchmarkMatrix, Generator, IntoBenchmarks, MeasurementSettings, DEFAULT_SETTINGS,
+};
+
+const SIZES: [usize; 14] = [
+    8, 16, 32, 64, 128, 256, 512, 1024, 2048, 4069, 8192, 16384, 32768, 65536,
+];
+
+struct Lcg(usize);
+
+impl Lcg {
+    fn next<T: TryFrom<usize>>(&mut self, max_value: usize) -> T {
+        self.0 = self.0.wrapping_mul(1664525).wrapping_add(1013904223);
+        T::try_from((self.0 >> 32) % max_value).ok().unwrap()
+    }
+}
+
+pub struct RandomCollection<C: FromSortedVec> {
+    rng: Lcg,
+    size: usize,
+    name: String,
+    value_dup_factor: usize,
+    phantom: PhantomData<C>,
+}
+
+impl<C: FromSortedVec> RandomCollection<C>
+where
+    C::Item: Ord + Copy + TryFrom<usize>,
+{
+    pub fn new(size: usize, value_dup_factor: usize) -> Self {
+        let type_name = type_name::<C::Item>();
+        let name = if value_dup_factor > 1 {
+            format!("{}/{}/dup-{}", type_name, size, value_dup_factor)
+        } else {
+            format!("{}/{}/nodup", type_name, size)
+        };
+
+        Self {
+            rng: Lcg(0),
+            size,
+            value_dup_factor,
+            name,
+            phantom: PhantomData,
+        }
+    }
+}
+
+impl<C: FromSortedVec> Generator for RandomCollection<C>
```

Comment:
I'm confused about why `haystack` and `needle` are first-class primitives in Tango — shouldn't it only be thinking in terms of "inputs" and "benchmarks"? If it also delineates between haystacks and needles, does that mean it doesn't generalize to other kinds of benchmarks without such concepts (e.g., factorials)? Alternatively, if it _does_ support "general" benchmarks, why have this specialization built into it as well?

Reply:
They are more of second-class primitives. For paired tests to be sensitive they need to operate on the same input. If the input is updated between iterations it should be updated at the same time between two algorithms. So harness need a way to control this process. This is how `Generator` trait was born.

Haystack/needle distinction is unusual, it's kind of "big" and "small" part of the input. Haystack is reused between benchmark samples, supposed to be costly in terms of generation (thus its generation time is not included in benchmark time). Needle is supposed to be cheap, generated for each iteration and its generation time is included in benchmark time.

This scheme can be generalized to arbitrary input, although it's not pretty. But it mirrors very important distinction which is present in almost any benchmarking harness including criterion. Under the hood we are almost never timing a single function call (iteration), but a bunch of them (sample). Some parts of the input can be cheaply generated per-iteration, another ones can not. This is needle/haystack separation.

I'm still unsure how easy it is to understand this concept. Maybe some another metaphor should be chosen, so I'm open to any feedback.